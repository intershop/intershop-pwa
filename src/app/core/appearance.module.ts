import { NgModule } from '@angular/core';
import { NgbPopoverConfig } from '@ng-bootstrap/ng-bootstrap';
import { ToastrModule } from 'ngx-toastr';

import { IconModule } from './icon.module';
import { ThemeService } from './utils/theme/theme.service';

@NgModule({
  imports: [
    IconModule,
    ToastrModule.forRoot({
      closeButton: true,
      timeOut: 3000,
      positionClass: 'toast-top-full-width', // toast-top-center
      preventDuplicates: true,
    }),
  ],
})
export class AppearanceModule {
  constructor(popoverConfig: NgbPopoverConfig, themeService: ThemeService) {
    popoverConfig.placement = 'top';
    popoverConfig.triggers = 'hover';
    popoverConfig.container = 'body';

    themeService.init();
  }
}
