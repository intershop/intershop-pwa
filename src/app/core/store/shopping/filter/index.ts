// API to access ngrx filter state
export * from './filter.actions';
export * from './filter.selectors';
