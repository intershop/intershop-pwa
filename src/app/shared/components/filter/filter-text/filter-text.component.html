<!--- TODO: implement Show All-->
<ul class="filter-list">
  <ng-container *ngFor="let facet of getFacets()">
    <li class="filter-item filter-layer{{ facet.level }}" [ngClass]="{ 'filter-selected': facet.selected }">
      <!-- selected -->
      <ng-container *ngIf="facet.selected; else notSelectedBox">
        <a (click)="filter(facet)" [attr.data-testing-id]="'filter-link-' + (facet.name | sanitize)">
          <span class="filter-item-name"> {{ facet.displayName }} </span>
          <span class="count"> ({{ facet.count }}) </span>
          <ng-container *ngIf="filterElement.displayType === 'text_clear'">
            <fa-icon [icon]="['fas', 'times']" class="float-right filter-clear"></fa-icon>
          </ng-container>
        </a>
      </ng-container>

      <!-- not selected -->
      <ng-template #notSelectedBox>
        <a
          class="filter-item-name"
          (click)="filter(facet)"
          [attr.data-testing-id]="'filter-link-' + (facet.name | sanitize)"
        >
          {{ facet.displayName }}
          <ng-container *ngIf="facet.level === maxLevel"> ({{ facet.count }}) </ng-container>
        </a>
      </ng-template>
    </li>
  </ng-container>
  <!-- show all/less -->
  <li class="filter-item" *ngIf="filterElement.limitCount !== -1 && facets.length > filterElement.limitCount">
    <a class="btn-link" (click)="showAll = !showAll">{{
      (showAll ? 'search.filter.show_less.link' : 'search.filter.show_all.link') | translate
    }}</a>
  </li>
</ul>
