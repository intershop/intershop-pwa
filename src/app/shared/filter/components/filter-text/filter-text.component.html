<!--- TODO: implement Show All-->
<ul class="filter-list">
  <ng-container *ngFor="let facet of facets">
    <li
      *ngIf="facet.displayName !== 'Show all'"
      class="filter-item filter-layer{{ facet.level }}"
      [ngClass]="{
        'filter-selected': facet.selected
      }"
    >
      <!-- selected -->
      <ng-container *ngIf="facet.selected && facet.level === maxLevel; else notSelectedBox">
        <!-- taxonomic category-filter (last level not unselectable) -->
        <ng-container *ngIf="filterElement.selectionType === 'taxonomic'; else textTemplate">
          <span class="filter-item-name"> {{ facet.displayName }} </span>
          <span class="count"> ({{ facet.count }}) </span>
        </ng-container>
        <ng-template #textTemplate>
          <a (click)="filter(facet)">
            <span class="filter-item-name"> {{ facet.displayName }} </span>
            <span class="count"> ({{ facet.count }}) </span>
            <ng-container *ngIf="filterElement.displayType === 'text_clear'">
              <fa-icon class="float-right filter-clear" [icon]="['fas', 'times']"></fa-icon>
            </ng-container>
          </a>
        </ng-template>
      </ng-container>

      <!-- not selected -->
      <ng-template #notSelectedBox>
        <a
          class="filter-item-name"
          (click)="filter(facet)"
          [attr.data-testing-id]="'filter-link-' + (facet.displayName | sanitize)"
        >
          {{ facet.displayName }}

          <ng-container *ngIf="facet.level === maxLevel"> ({{ facet.count }}) </ng-container>
        </a>
      </ng-template>
    </li>
  </ng-container>
</ul>
