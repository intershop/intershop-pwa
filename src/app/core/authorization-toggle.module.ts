import { ModuleWithProviders, NgModule } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';

import { AuthorizationToggleDirective } from './directives/authorization-toggle.directive';
import { AuthorizationToggleService, checkPermission } from './utils/authorization-toggle/authorization-toggle.service';
import { whenTruthy } from './utils/operators';

@NgModule({
  declarations: [AuthorizationToggleDirective],
  exports: [AuthorizationToggleDirective],
})
export class AuthorizationToggleModule {
  private static permissions = new ReplaySubject<string[]>(1);

  static forTesting(...permissions: string[]): ModuleWithProviders<AuthorizationToggleModule> {
    AuthorizationToggleModule.switchTestingPermissions(...permissions);
    return {
      ngModule: AuthorizationToggleModule,
      providers: [
        {
          provide: AuthorizationToggleService,
          useValue: {
            isAuthorizedTo: (permission: string) =>
              AuthorizationToggleModule.permissions.pipe(
                whenTruthy(),
                map(perms => checkPermission(perms, permission))
              ),
          },
        },
      ],
    };
  }

  static switchTestingPermissions(...permissions: string[]) {
    AuthorizationToggleModule.permissions.next(permissions);
  }
}

export { AuthorizationToggleService } from './utils/authorization-toggle/authorization-toggle.service';
export { AuthorizationToggleGuard } from './guards/authorization-toggle.guard';
