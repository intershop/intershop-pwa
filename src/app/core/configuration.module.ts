import { NgModule } from '@angular/core';

import { SPECIAL_HTTP_ERROR_HANDLER } from './interceptors/icm-error-mapper.interceptor';
import { createPaymentErrorHandler } from './utils/http-error/create-payment.error-handler';
import { editPasswordErrorHandler } from './utils/http-error/edit-password.error-handler';
import { LoginUserErrorHandler } from './utils/http-error/login-user.error-handler';
import { requestReminderErrorHandler } from './utils/http-error/request-reminder.error-handler';
import { updatePasswordErrorHandler } from './utils/http-error/update-password.error-handler';

@NgModule({
  providers: [
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: updatePasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useClass: LoginUserErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: requestReminderErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: editPasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: createPaymentErrorHandler, multi: true },
  ],
})
export class ConfigurationModule {}
