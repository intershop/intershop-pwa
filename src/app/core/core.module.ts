import { APP_BASE_HREF, PlatformLocation } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { ErrorHandler, NgModule, Optional, SkipSelf } from '@angular/core';
import { ServiceWorkerModule } from '@angular/service-worker';
import { TranslateModule } from '@ngx-translate/core';

import { AppearanceModule } from './appearance.module';
import { ConfigurationModule } from './configuration.module';
import { FeatureToggleModule } from './feature-toggle.module';
import { IdentityProviderModule } from './identity-provider.module';
import { ICMErrorMapperInterceptor } from './interceptors/icm-error-mapper.interceptor';
import { IdentityProviderInterceptor } from './interceptors/identity-provider.interceptor';
import { MockInterceptor } from './interceptors/mock.interceptor';
import { InternationalizationModule } from './internationalization.module';
import { StateManagementModule } from './state-management.module';
import { DefaultErrorHandler } from './utils/default-error-handler';

@NgModule({
  imports: [
    AppearanceModule,
    ConfigurationModule,
    FeatureToggleModule,
    HttpClientModule,
    IdentityProviderModule,
    InternationalizationModule,
    ServiceWorkerModule.register('ngsw-worker.js', { enabled: SERVICE_WORKER }),
    StateManagementModule,
  ],
  providers: [
    { provide: HTTP_INTERCEPTORS, useClass: ICMErrorMapperInterceptor, multi: true },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: IdentityProviderInterceptor,
      multi: true,
    },
    { provide: HTTP_INTERCEPTORS, useClass: MockInterceptor, multi: true },
    { provide: ErrorHandler, useClass: DefaultErrorHandler },
    {
      provide: APP_BASE_HREF,
      useFactory: (s: PlatformLocation, baseHref: string) => baseHref || s.getBaseHrefFromDOM(),
      deps: [PlatformLocation, [new Optional(), new SkipSelf(), APP_BASE_HREF]],
    },
  ],
  // exports needed to use the cookie banner in the AppComponent
  exports: [TranslateModule],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error('CoreModule is already loaded. Import it in the AppModule only');
    }
  }
}
