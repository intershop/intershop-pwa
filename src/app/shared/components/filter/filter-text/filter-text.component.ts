import { ChangeDetectionStrategy, Component, EventEmitter, Input, OnInit, Output } from '@angular/core';

import { Facet } from 'ish-core/models/facet/facet.model';
import { Filter } from 'ish-core/models/filter/filter.model';
import { URLFormParams } from 'ish-core/utils/url-form-params';

/**
 * The Filter Text Component displays a filter group. The facets of the filter group are presented as links with optional clear-button.
 *
 * @example
 * <ish-filter-text
 *               [filterElement]="element"
 *               (applyFilter)="applyFilter($event)"
 * </ish-filter-text>
 */
@Component({
  selector: 'ish-filter-text',
  templateUrl: './filter-text.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FilterTextComponent implements OnInit {
  @Input() filterElement: Filter;
  @Output() applyFilter: EventEmitter<{ searchParameter: URLFormParams }> = new EventEmitter();

  /**
   * two-way-binding (banana in a box) [(showAll)]="showAllElements[element.name]"
   */
  @Output()
  showAllChange = new EventEmitter<boolean>();
  private showAllValue = false;
  @Input()
  get showAll() {
    return this.showAllValue;
  }
  set showAll(val) {
    this.showAllValue = val;
    this.showAllChange.emit(this.showAllValue);
  }

  maxLevel = 0;
  facets: Facet[] = [];

  ngOnInit() {
    this.maxLevel =
      Math.max.apply(
        Math,
        this.filterElement.facets.map(o => o.level)
      ) || 0;
    this.facets = this.filterElement.facets.filter(x => x.selected || !this.maxLevel || x.level >= this.maxLevel);
  }

  filter(facet: Facet) {
    this.applyFilter.emit({ searchParameter: facet.searchParameter });
  }

  /**
   * sort selected to top, increase limitCount to selectedCount on showLess
   */
  getFacets() {
    const facets = [...this.facets];

    if (this.showAll || this.maxLevel >= 1 || this.filterElement.limitCount === -1) {
      return facets;
    }

    const selectedFacetsCount = facets.filter(x => x.selected).length;

    return facets
      .sort((a, b) => (a.selected > b.selected ? -1 : a.selected < b.selected ? 1 : 0))
      .slice(0, Math.max(this.filterElement.limitCount || 0, selectedFacetsCount));
  }
}
