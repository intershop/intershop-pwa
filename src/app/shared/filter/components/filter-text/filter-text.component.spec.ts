import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { FaIconComponent } from '@fortawesome/angular-fontawesome';
import { MockComponent } from 'ng-mocks';

import { Filter } from 'ish-core/models/filter/filter.model';
import { SanitizePipe } from 'ish-core/pipes/sanitize.pipe';

import { FilterTextComponent } from './filter-text.component';

describe('Filter Text Component', () => {
  let component: FilterTextComponent;
  let fixture: ComponentFixture<FilterTextComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [FilterTextComponent, MockComponent(FaIconComponent), SanitizePipe],
    }).compileComponents();
  }));

  beforeEach(() => {
    const filterElement = {
      name: 'Brands',
      facets: [
        { name: 'AsusName', count: 4, displayName: 'Asus' },
        { name: 'LogitechName', count: 5, displayName: 'Logitech', selected: true },
      ],
    } as Filter;
    fixture = TestBed.createComponent(FilterTextComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    component.filterElement = filterElement;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });
});
