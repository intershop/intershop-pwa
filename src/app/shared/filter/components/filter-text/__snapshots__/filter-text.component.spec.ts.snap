// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Text Component should be created 1`] = `
<ul class="filter-list">
  <li class="filter-item filter-layer">
    <a class="filter-item-name" data-testing-id="filter-link-Asus"> Asus </a>
  </li>
  <li class="filter-item filter-layer filter-selected">
    <a class="filter-item-name" data-testing-id="filter-link-Logitech"> Logitech </a>
  </li>
</ul>
`;
